scripts/documentation/tsdoc.ts

```diff
@@ -0,0 +1,284 @@
+import {Documentalist, TypescriptPlugin} from '@documentalist/compiler';
+import * as fs from 'fs';
+
+const markdownHeaderLevel = '###';
+const markdownSubHeaderLevel = '####';
+
+class TypeScriptDocumentationGenerator {
+  markdownBuffer: {};
+  docData?: {};
+
+  constructor() {
+    this.docData = {};
+    this.markdownBuffer = {};
+  }
+
+  /**
+   * Generates JSON from source files TypeScript documentation.
+   * This contains all the esmodule classes (ie. foundations, adapters, components) in JSON format.
+   * @returns Promise(json)
+   */
+  generateJSONFromFiles() {
+    return new Promise((resolve, reject) => {
+      new Documentalist()
+        .use(/\.ts$/, new TypescriptPlugin({
+          excludePaths: ['node_modules'],
+          includeDeclarations: true,
+        }))
+        .documentGlobs('packages/**/*') // ← async operation, returns a Promise
+        .then((docs) => {
+          resolve(docs);
+          // TODO - remove when all docs are complete
+          return JSON.stringify(docs, null, 2);
+        })
+        .then((json) => {
+          // TODO: comment out - just for debugging purposes.
+          fs.writeFileSync('docs.json', json);
+        })
+        .catch((error) => console.error(error)); // tslint:disable-line
+    });
+  }
+
+  /**
+   * The main function of this class. Iterates through all modules found in docData (json).
+   * This should already be generated from this.generateJSONFromFiles().
+   * @param docData json containing documentation from documentalist
+   */
+  generateDocs(docData) {
+    this.docData = docData.typescript;
+    Object.keys(this.docData).forEach((module) => {
+      console.log(`-- generating docs for ${module}`); // tslint:disable-line
+      this.generateDocsForModule(module);
+    });
+
+    this.generateMarkdownFileFromBuffer();
+  }
+
+  /**
+   * Creates documentation for a specified `esmodule`.
+   * @param esmodule module name (ie. MDCSelectIconFoundation)
+   */
+  generateDocsForModule(esmodule: string) {
+    const moduleName = `${markdownHeaderLevel} ${esmodule}\n\n`;
+    const markdownString =
```

Comment:
I'm wary about concatenating strings, this makes it hard to reformat the structure later. We need a templating engine, something like Nunjucks. This makes it easier to restructure the markdown output without looking at the code.

Reply:
Ok I will look into Nunjucks